src/colmap/feature/pairing.cc

```diff
@@ -639,51 +639,105 @@ Eigen::RowMajorMatrixXf SpatialPairGenerator::ReadPositionPriorData(
+          position_prior(2) == 0) {
+        continue;
+      }
     }
 
+    int position_idx = position_idxs_.size();
     position_idxs_.push_back(i);
 
+    if (position_idx == 0) {
+      switch (pose_prior->coordinate_system) {
+        case PosePrior::CoordinateSystem::WGS84: {
+          ells[0](0) = position_prior(0);
+          ells[0](1) = position_prior(1);
+          ells[0](2) = options_.ignore_z ? 0 : position_prior(2);
+
+          const auto xyzs = gps_transform.EllipsoidToECEF(ells);
+          position_offset = xyzs[0];
+        } break;
+        case PosePrior::CoordinateSystem::UNDEFINED:
+        default:
+        case PosePrior::CoordinateSystem::CARTESIAN:
+          position_offset(0) = position_prior(0);
+          position_offset(1) = position_prior(1);
+          position_offset(2) = (options_.ignore_z ? 0 : position_prior(2));
+      }
+      position_matrix(position_idx, 0) = 0.0f;
+      position_matrix(position_idx, 1) = 0.0f;
+      position_matrix(position_idx, 2) = 0.0f;
+      continue;
+    }
+
     switch (pose_prior->coordinate_system) {
       case PosePrior::CoordinateSystem::WGS84: {
         ells[0](0) = position_prior(0);
         ells[0](1) = position_prior(1);
         ells[0](2) = options_.ignore_z ? 0 : position_prior(2);
 
         const auto xyzs = gps_transform.EllipsoidToECEF(ells);
-        position_matrix(num_positions, 0) = static_cast<float>(xyzs[0](0));
-        position_matrix(num_positions, 1) = static_cast<float>(xyzs[0](1));
-        position_matrix(num_positions, 2) = static_cast<float>(xyzs[0](2));
+        position_matrix(position_idx, 0) =
+            static_cast<float>(xyzs[0](0) - position_offset(0));
+        position_matrix(position_idx, 1) =
+            static_cast<float>(xyzs[0](1) - position_offset(1));
+        position_matrix(position_idx, 2) =
+            static_cast<float>(xyzs[0](2) - position_offset(2));
       } break;
       case PosePrior::CoordinateSystem::UNDEFINED:
       default:
         LOG(WARNING) << "Unknown coordinate system for image " << image_ids_[i]
                      << ", assuming cartesian.";
       case PosePrior::CoordinateSystem::CARTESIAN:
-        position_matrix(num_positions, 0) =
-            static_cast<float>(position_prior(0));
-        position_matrix(num_positions, 1) =
-            static_cast<float>(position_prior(1));
-        position_matrix(num_positions, 2) =
-            static_cast<float>(options_.ignore_z ? 0 : position_prior(2));
+        position_matrix(position_idx, 0) =
+            static_cast<float>(position_prior(0) - position_offset(0));
+        position_matrix(position_idx, 1) =
+            static_cast<float>(position_prior(1) - position_offset(1));
+        position_matrix(position_idx, 2) = static_cast<float>(
+            options_.ignore_z ? 0 : position_prior(2) - position_offset(2));
     }
-
-    num_positions += 1;
   }
   return position_matrix;
```

Comment:
Wouldn't it be easier to `return position_matrix.topRows(num_positions);` instead of the proposed fix?